### PR TITLE
Implement `flake8-newspaper-style`

### LINKED ISSUE #2436: Implement `flake8-newspaper-style`
https://pypi.org/project/flake8-newspaper-style/
Rule to check for logic ordering of function calls within a file.

### PR DESCRIPTION
**Pull request by sasanjac**
Fixes #2436

### CODE REVIEW

**Comment by JonathanPlasse** [comment]
What happens if two functions or classes call each other?

**Comment by sasanjac** [comment]
`RecursionError` :D
Yeah, I'm going to have to think about solving that.

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:44]: Nit: `checker.ctx.bindngs[*index]` (you already have the index)

**Review by MichaReiser — APPROVED** [review]
Looks good. Thanks for working on this. Let's add a test case for an invalid caller name and one for recursion (although I don't see any issues).

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:27]: oops 🙈

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:42]: 1. `bindings` can't miss `caller_name`.
2. If `bindings` misses `callee_name`, we get a panic, however, when we instead get the max. int, this conditions always evaluates to false and thus no violation is reported.

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:44]: Where? I can't find it 🙈

**Comment by sasanjac** [comment]
> What happens if two functions or classes call each other?

Should be fixed now

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:9]: I think this may need to be rebased, some of these got moved around the APIs changed a bit today.

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:43]: Nit: don't forget to remove :)

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:62]: Do you mind documenting the various cases here? I'm having a little trouble following from the code alone (e.g., which path is intended for which combination of definitions).

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:9]: Yeah I'm already on it, however  it's taking me some time to understand some patterns in Rust 😅

Comment (MichaReiser) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:9]: Let me know if I can help you with the rebase or if you have any questions

Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs]: Thanks :) I think I got it now but maybe you could check my new changes to see if I understood correctly

File: crates/ruff_python_ast/src/context.rs
Comment (MichaReiser) [inline crates/ruff_python_ast/src/context.rs]: I don't think this returns the `grandparent` scope

* `scope_stack.iter()` returns the reversed stack
* `next` returns the first element 

-> That would mean it returns the global scope. 

I believe this should instead be

```rust
self.scope_stack.iter().nth(2).map(|index| &self.scopes[*index])
```

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:134]: Does this logic recursively traverse the body of the function:

```python
def a():
	if True:  # child of b's body
		b()


def b(): 
  if False:
	a()
```

File: crates/ruff_python_ast/src/context.rs
Comment (sasanjac) [reply crates/ruff_python_ast/src/context.rs]: Yeah, it used to be like this but now `.nth(2)` returns the function scope (current scope) and `.nth(0)` or `.next()` returns the module scope. Or am I thinking wrong: `function < class < module <=> current < parent < grandparent`

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:134]: I don't think yet but it definitely should at least until it first encounters a mutual recursion. I'm going to fix this

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:69]: Should we be validating here that this function is a method (instance or class), and that this call is on `self` or `cls` as appropriate?

E.g., these are false positives:

```py
class A:
    def foo(self):
        return 1

    def bar(self):
        bop.foo()  # NEW100
        b = bop.foo()  # NEW100
        return self.bar(b)
```

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:118]: Hmm, if I understand what's going on here, I consider this an implementation detail. It's not guaranteed that these bindings are increasing as we parse through the program. Can we instead look at the `source` field on `Binding`, and compare based on location?

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:29]: Nit: this isn't necessary, right? Since we checked this a level above?

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:49]: Should we be catching nested functions? Like this?

```py
def foo():

    def bop():
        bar()

    def bar():
        x = 1

    def baz():
        x = 1
```

Right now, neither this implementation nor `flake8-newspaper-style` catch this.

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:147]: Nit: use `Stmt` in lieu of `Located<StmtKind>`, and `&str` in lieu of `&String`, if possible?

**Review by charliermarsh — COMMENTED** [review]
Thanks, really sorry for the double review, the comments helped me understand the code better so have one or two more questions!

**Comment by charliermarsh** [comment]
I'd love to get the ecosystem CI check running on this, I wonder why it's not appearing.

**Comment by MichaReiser** [comment]
> I'd love to get the ecosystem CI check running on this, I wonder why it's not appearing.

The diff is too large and exceeds the GitHub summary... https://github.com/charliermarsh/ruff/actions/runs/4457214954/jobs/7828206645?pr=3561

File: crates/ruff_python_ast/src/context.rs
Comment (MichaReiser) [inline crates/ruff_python_ast/src/context.rs:303]: We should investigate why `nth(2)` doesn't give you the expected result. I recommend moving this function to your implementation if it is because of the way you traverse the AST.

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs:131]: Is it intended that the rule uses bitwise and?

```suggestion
                if matches!(kind, BindingKind::FunctionDefinition)
                    && !mutual_recursion(body, &caller_name.to_string())
```

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs]: For now, it does only find mutual recursion using two methods. However we could remove that altogether and just require people to ignore the violation with `nova` when a recursion does indeed occur.

**Comment by sasanjac** [comment]
> Thanks, really sorry for the double review, the comments helped me understand the code better so have one or two more questions!

No thanks for your comments, it made me understand a bit more about the projects structure and Rust in general

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs]: No 🙈, fixed

File: crates/ruff_python_ast/src/context.rs
Comment (sasanjac) [reply crates/ruff_python_ast/src/context.rs]: I modified it and instead get a copy of the scope stack which is the popped/pushed according to the scope.

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs]: Fixed

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs]: Done, however we might have to think about how it is possible to completely generalize this problem. E.g. catching nested functions is done recursively, so the nested level shouldn't matter but nested classes are not caught atm.

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs]: Fixed

File: crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs
Comment (sasanjac) [reply crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs]: Yes, fixed

**Comment by sasanjac** [comment]
@charliermarsh @MichaReiser  Is there anything I can do to advance this PR?

**Comment by charliermarsh** [comment]
@sasanjac - No, really sorry, this is just on me/us to define a formal policy around which plugins we include in Ruff or don't, and why -- which we're working on. Some plugins are obvious or easy to include, but in this case I'm moving slowly because it's a very opinionated rule that will trigger a lot of violations on any project using `"ALL"` (which is discouraged, but does have usages out there), and so I want to have a clear decision-making process in place before we merge this.

I'm happy to be responsible for rebasing etc. though I again am having some trouble pushing to this branch, maybe because of the backticks, I'm not sure.

**Comment by sasanjac** [comment]
> I'm happy to be responsible for rebasing etc. though I again am having some trouble pushing to this branch, maybe because of the backticks, I'm not sure.

That's really strange. I can't edit any settings on this PR and afaik edits are allowed for maintainers by default. Did you check out the branch in my fork?

**Comment by charliermarsh** [comment]
Thank you again for all your work here. I'm sorry to have let this one linger, but... I've made the decision to close it out. I don't think this plugin has the right characteristics to be part of Ruff right now -- it'd be one of the least-downloaded existing plugins we've implemented (~5k downloads per month), but with a high flag rate, and a complex interface + implementation. It would also conflict with `ssort`, which there's been interest in implementing, and has about 10x the downloads in the wild. None of these characteristics would be individually disqualifying, but as a whole, I don't think it makes sense to merge.

I apologize for not making this call sooner, _especially_ before you'd started on the implementation -- I'm trying to get better at making those kinds of decisions earlier in the development process, but am still growing into my responsibilities as a maintainer.

**Pull request closed without merging**
